keps/sig-cli/0015-kubectl-plugins.md

```diff
@@ -0,0 +1,151 @@
+* Relay all information given to `kubectl` (via command line args) to plugins as-is.
+  Plugins receive all arguments and flags provided by users and are responsible for adjusting their behavior
+  accordingly.
+* Provide a way for plugins to override specific subcommands, while preventing other paths from being overriden.
+* Provide a standard package of common cli utilities that is consumed by `kubectl` and plugins alike.
+  These utilities would provide all of the major pieces needed to build a standard cli command, such as retrieving 
+  REST configuration, REST clients, printing, etc. (See https://github.com/kubernetes/features/issues/581).
+
+### Non-Goals
+
+* The new plugin mechanism will not be a "plugin installer" or wizard. It will not have specific or baked-in knowledge 
+  regarding a plugin's location or composition, nor will it it provide a way to download or unpack plugins in a correct 
+  location.
+* Plugin discovery is not a main focus of this mechanism. As such, it will not attempt to collect data about every 
+  plugin that exists in an environment.
+* Plugin management is out of the scope of this design. A mechanism for updating and managing lifecycle of existing 
+  plugins should be covered as a separate design (See https://github.com/kubernetes/community/pull/2340).
+
+## Proposal
+
+### Scenarios
+
+* Developer wants to create and expose a plugin to `kubectl`.
+  They use a programming language of their choice and create an executable file.
+  The executable's filename consists of the command path to replace, and is prefixed with `kubectl-`.
+  The executable file is placed on the user's PATH, preferably under `/usr/local/bin`.
+  If a plugin is written using a dynamic language, and depends on logic defined in additional files, those
+  files are placed elsewhere in the user's system (e.g. ~/.kubeplugins/myplugin/helper1.py) and not in the user's `PATH`.
+* User wants to "install" a plugin `kubectl-plugin-foo`
+  They move the executable file `kubectl-plugin-foo` to a location on their path, and access the plugin via the command `kubectl plugin foo`.
+* User wants a subset of their plugin executables to live in a specific location on their filesystem.
+  The user creates a directory in the desired location (`~/Documents/foo-plugins`) and adds that location to their `PATH`.
+
+### Implementation Details/Notes/Constraints
+
+The current implementation details for the proposed design rely on using a plugin executable's name to determine what
+command the plugin is adding or overwriting.
+For a given command `kubectl plugin foo --bar baz`, an executable `kubectl-plugin-foo` will be matched on a user's `PATH`,
+and the arguments `--bar baz` will be passed to it in that order.
+
+A potential limitation of this could present itself in the order of arguments provided by a user.
+A user could intend to run a plugin `kubectl-plugin-foo-bar` with the flag `--baz` with the following command
```

Comment:
Arguments are not under consideration when matching plugin name,  so this is not a limitation here.

Reply:
I would say the limitation would still exist in the constraint that is placed in the order in which a user can provide subcommands and flags. Although the plugin mechanism does not handle or parse any args meant for a plugin, the plugin itself might not be matched if a --flag is specified before its full path